pkg/engine/datasource/graphql_datasource/graphql_subscription_client.go

```diff
@@ -132,19 +134,37 @@ func (c *WebSocketGraphQLSubscriptionClient) Subscribe(ctx context.Context, opti
 	if err != nil {
 		return err
 	}
-	msgType, connectionAckMsg, err := conn.Read(ctx)
-	if err != nil {
-		return err
-	}
-	if msgType != websocket.MessageText {
-		return fmt.Errorf("unexpected msg type")
-	}
-	connectionAck, err := jsonparser.GetString(connectionAckMsg, "type")
-	if err != nil {
-		return err
-	}
-	if connectionAck != "connection_ack" {
-		return fmt.Errorf("expected connection_ack, got: %s", connectionAck)
+
+	// wait for ack and/or ka
+	timer := time.NewTimer(ackWaitTimeout)
+	for {
+		select {
+		case <-timer.C:
+			return fmt.Errorf("timeout while waiting for connection_ack")
+		default:
+		}
+
+		msgType, msg, err := conn.Read(ctx)
+		if err != nil {
+			return err
+		}
+		if msgType != websocket.MessageText {
+			return fmt.Errorf("unexpected message type")
+		}
+
+		respType, err := jsonparser.GetString(msg, "type")
+		if err != nil {
+			return err
+		}
+
+		if respType != "ka" && respType != "connection_ack" {
+			return fmt.Errorf("expected connection_ack or ka, got %s", respType)
+		}
+		if respType == "ka" {
+			continue
+		} else if respType == "connection_ack" {
```

Comment:
It's trivial, but you can write this `if/else if` block with a switch-case statement. It's more idiomatic.

Reply:
yeah with the switch a label for the break would need to be used to point it to the for loop rather than the switch https://github.com/TykTechnologies/graphql-go-tools/pull/277#discussion_r913431256